File: app/src/main/java/com/soutvoid/gamesproject/ui/screen/queryEdit/QueryEditActivityView.java
package com.soutvoid.gamesproject.ui.screen.queryEdit;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.support.v4.content.ContextCompat;
import android.widget.Switch;
import android.widget.TextView;

import com.agna.ferro.mvp.component.ScreenComponent;
import com.annimon.stream.Stream;
import com.google.android.flexbox.FlexboxLayout;
import com.soutvoid.gamesproject.domain.genre.Genre;
import com.soutvoid.gamesproject.ui.base.activity.BaseActivityView;
import com.soutvoid.gamesproject.ui.base.activity.BasePresenter;
import com.soutvoid.gamesproject.ui.base.widgets.IgdbToolbar;
import com.soutvoid.gamesproject.ui.screen.queryEdit.data.QueryData;
import com.soutvoid.gamesproject.ui.screen.queryEdit.widgets.choosableDate.ChoosableDateTextView;
import com.soutvoid.gamesproject.ui.screen.queryEdit.widgets.genreToggleBtn.GenreToggleBtn;
import com.wdullaer.materialdatetimepicker.date.MonthAdapter;

import java.util.List;

import javax.inject.Inject;

import butterknife.BindView;
import soutvoid.com.gamesproject.R;

public class QueryEditActivityView extends BaseActivityView {

    @BindView(R.id.toolbar)
    IgdbToolbar toolbar;
    @BindView(R.id.section_released_from_date)
    ChoosableDateTextView releasedFromDate;
    @BindView(R.id.section_released_to_date)
    ChoosableDateTextView releasedToDate;
    @BindView(R.id.section_released_include_to)
    Switch includeTo;

    @BindView(R.id.section_released_to)
    TextView releasedTo;

    @BindView(R.id.section_genres_flexbox)
    FlexboxLayout genresFlexbox;

    @Inject
    QueryEditActivityPresenter presenter;

    public static void start(Context context) {
        Intent intent = new Intent(context, QueryEditActivityView.class);
        context.startActivity(intent);
    }

    @Override
    public BasePresenter getPresenter() {
        return presenter;
    }

    @Override
    public String getName() {
        return getClass().getSimpleName();
    }

    @Override
    protected int getContentView() {
        return R.layout.activity_query_edit;
    }

    @Override
    protected ScreenComponent createScreenComponent() {
        return DaggerQueryEditActivityComponent.builder()
                .activityModule(getActivityModule())
                .appComponent(getAppComponent())
                .build();
    }

    @Override
    protected void onCreate(Bundle savedInstanceState, boolean viewRecreated) {
        super.onCreate(savedInstanceState, viewRecreated);

        setupToolbar();
        setupDatePickers();
        setupViews();
    }

    private void setupToolbar() {
        toolbar.showExpandedPurpleToolbarSaveInput(
                getString(R.string.new_query),
                this,
                v -> presenter.onSaveClick());
    }

    private void setupDatePickers() {
        releasedFromDate.setActivity(this);
        releasedToDate.setActivity(this);
    }

    private void setupViews() {
        includeTo.setOnCheckedChangeListener((btn, isChecked) ->
                presenter.onReleasedIncludeToClicked(btn, isChecked));
    }


    String getNameInput() {
        return toolbar.getInputText();
    }

    void setNameInput(String input) {
        toolbar.setInputText(input);
    }

    MonthAdapter.CalendarDay getReleasedFromDate() {
        return releasedFromDate.getSelectedDay();
    }

    MonthAdapter.CalendarDay getReleasedToDate() {
        return releasedToDate.getSelectedDay();
    }

    void setGenresList(List<Genre> genres) {
        Stream.of(genres).forEach(
                genre -> {
                    GenreToggleBtn genreToggleBtn = new GenreToggleBtn(this);
                    genreToggleBtn.setText(genre.getName());
                    genresFlexbox.addView(genreToggleBtn);
                }
        );
    }

    QueryData getData() {
        return new QueryData(
                getNameInput(),
                getReleasedFromDate(),
                getReleasedToDate()
        );
    }

    void toggleReleasedToDatePicker(boolean enabled) {
        releasedToDate.setEnabled(enabled);
        int textColorId = R.color.black_transparent_40;
        if (enabled)
            textColorId = R.color.colorTextInverse;
        releasedTo.setTextColor(ContextCompat.getColor(this, textColorId));
        releasedToDate.setTextColor(ContextCompat.getColor(this, textColorId));
    }

    boolean isReleasedToIncluded() {
        return includeTo.isChecked();
    }
}

File: app/src/main/java/com/soutvoid/gamesproject/ui/screen/queryEdit/QueryEditActivityView.kt
package com.soutvoid.gamesproject.ui.screen.queryEdit

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.support.v4.content.ContextCompat
import android.widget.Switch
import android.widget.TextView
import butterknife.BindView
import com.agna.ferro.mvp.component.ScreenComponent
import com.annimon.stream.Stream
import com.google.android.flexbox.FlexboxLayout
import com.soutvoid.gamesproject.domain.genre.Genre
import com.soutvoid.gamesproject.ui.base.activity.BaseActivityView
import com.soutvoid.gamesproject.ui.base.activity.BasePresenter
import com.soutvoid.gamesproject.ui.base.widgets.IgdbToolbar
import com.soutvoid.gamesproject.ui.screen.queryEdit.data.QueryData
import com.soutvoid.gamesproject.ui.screen.queryEdit.widgets.choosableDate.ChoosableDateTextView
import com.soutvoid.gamesproject.ui.screen.queryEdit.widgets.genreToggleBtn.GenreToggleBtn
import com.wdullaer.materialdatetimepicker.date.MonthAdapter
import soutvoid.com.gamesproject.R
import java.util.*
import javax.inject.Inject

class QueryEditActivityView : BaseActivityView() {

    @BindView(R.id.toolbar)
    var toolbar: IgdbToolbar? = null
    @BindView(R.id.section_released_from_date)
    var releasedFromDate: ChoosableDateTextView? = null
    @BindView(R.id.section_released_to_date)
    var releasedToDate: ChoosableDateTextView? = null
    @BindView(R.id.section_released_include_to)
    var includeTo: Switch? = null

    @BindView(R.id.section_released_to)
    var releasedTo: TextView? = null

    @BindView(R.id.section_genres_flexbox)
    var genresFlexbox: FlexboxLayout? = null

    @Inject
    lateinit internal var presenter: QueryEditActivityPresenter

    override fun getPresenter(): BasePresenter<*> {
        return presenter
    }

    override fun getName(): String {
        return javaClass.simpleName
    }

    override fun getContentView(): Int {
        return R.layout.activity_query_edit
    }

    override fun createScreenComponent(): ScreenComponent<*> {
        return DaggerQueryEditActivityComponent.builder()
                .activityModule(activityModule)
                .appComponent(appComponent)
                .build()
    }

    override fun onCreate(savedInstanceState: Bundle?, viewRecreated: Boolean) {
        super.onCreate(savedInstanceState, viewRecreated)

        setupToolbar()
        setupDatePickers()
        setupViews()
    }

    private fun setupToolbar() {
        toolbar!!.showExpandedPurpleToolbarSaveInput(
                getString(R.string.new_query),
                this
        ) { v -> presenter!!.onSaveClick() }
    }

    private fun setupDatePickers() {
        releasedFromDate!!.setActivity(this)
        releasedToDate!!.setActivity(this)
    }

    private fun setupViews() {
        includeTo!!.setOnCheckedChangeListener { btn, isChecked -> presenter!!.onReleasedIncludeToClicked(btn, isChecked) }
    }


    internal var nameInput: String
        get() = toolbar!!.inputText
        set(input) {
            toolbar!!.inputText = input
        }

    internal fun getReleasedFromDate(): MonthAdapter.CalendarDay {
        return releasedFromDate!!.selectedDay
    }

    internal fun getReleasedToDate(): MonthAdapter.CalendarDay {
        return releasedToDate!!.selectedDay
    }

    internal fun setGenresList(genres: List<Genre>) {
        Stream.of(genres).forEach { genre ->
            val genreToggleBtn = GenreToggleBtn(this)
            genreToggleBtn.text = genre.name
            genresFlexbox!!.addView(genreToggleBtn)
        }
    }

    internal val data: QueryData
        get() = QueryData(
                nameInput,
                getReleasedFromDate(),
                getReleasedToDate(),
                selectedGenres
        )

    internal fun toggleReleasedToDatePicker(enabled: Boolean) {
        releasedToDate!!.isEnabled = enabled
        var textColorId = R.color.black_transparent_40
        if (enabled)
            textColorId = R.color.colorTextInverse
        releasedTo!!.setTextColor(ContextCompat.getColor(this, textColorId))
        releasedToDate!!.setTextColor(ContextCompat.getColor(this, textColorId))
    }

    internal val isReleasedToIncluded: Boolean
        get() = includeTo!!.isChecked

    internal val selectedGenres: List<Int>
        get() {
            val genres = ArrayList<Int>()
            for (i in 0..genresFlexbox!!.childCount - 1) {
                if ((genresFlexbox!!.getChildAt(i) as GenreToggleBtn).isChecked)
                    genres.add(i)
            }
            return genres
        }

    companion object {

        fun start(context: Context) {
            val intent = Intent(context, QueryEditActivityView::class.java)
            context.startActivity(intent)
        }
    }
}
